File: algorithms/397. Integer Replacement/solution.py
class Solution(object):
    def integerReplacement(self, n):
        """
        :type n: int
        :rtype: int
        """
        count = 0
        while n > 1:
            if n & 1 == 0:
                n /= 2
            elif n % 4 == 1 or n == 3:
                n -= 1
            else:
                n += 1
            count += 1
        return count

File: algorithms/397. Integer Replacement/solution.cc
class Solution {
public:
    int integerReplacement(int n) {
        int count = 0;
        unsigned int m = n; // for 2147483647
        while ( m > 1 ) {
            if ( m % 2 == 0 )
                m /= 2;
            else if ( m % 4 == 1 || 3 == m )
                --m;
            else
                ++m;
            ++count;
        }
        return count;
    }
};
